src/colmap/scene/imu.h

```diff
@@ -0,0 +1,131 @@
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of ETH Zurich and UNC Chapel Hill nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include "colmap/geometry/rigid3.h"
+#include "colmap/sensor/imu.h"
+#include "colmap/util/eigen_alignment.h"
+#include "colmap/util/logging.h"
+#include "colmap/util/types.h"
+
+#include <vector>
+
+#include <Eigen/Dense>
+#include <Eigen/Geometry>
+
+namespace colmap {
+
+// An Imu class storing the sensor information and a linked visual camera
+// TODO: support rigs + unify sensors
+class Imu {
+ public:
+  Imu() = default;
+  ~Imu() = default;
+  ImuCalibration calib;
+  camera_t imu_id = kInvalidCameraId;
+
+  // information for the associated visual camera. TODO: change to rigs
+  camera_t camera_id = kInvalidCameraId;  // the camera linked to IMU.
+  Rigid3d cam_to_imu;
+};
+
+// A state class storing speed and biases for discrete-time optimization
+class ImuState {
+ public:
+  ImuState() = default;
+  ~ImuState() = default;
+  inline const double* Data() const;
+
+  inline void SetVelocity(const Eigen::Vector3d& vec);
+  inline const Eigen::Vector3d Velocity() const;
+  inline const double* VelocityPtr();
+
+  inline void SetAccBias(const Eigen::Vector3d& vec);
+  inline const Eigen::Vector3d AccBias() const;
+  inline const double* AccBiasPtr();
+
+  inline void SetGyroBias(const Eigen::Vector3d& vec);
+  inline const Eigen::Vector3d GyroBias() const;
+  inline const double* GyroBiasPtr();
+
+  inline const camera_t& ImuId() const;
+  inline camera_t ImuId();
+  inline const image_t& ImageId() const;
+  inline image_t ImageId();
+
+ private:
+  Eigen::Matrix<double, 9, 1> data_ =
+      Eigen::Matrix<double, 9, 1>::Zero();  // 3-DoF speed + 6-DoF biases (acc +
+                                            // gyro)
```

Comment:
Could be nicer to store them here as three independent Vector3d's and then, whenever we need them as a singl state-vector (e.g., during optimization), copy them into a Vector9d ?

Then this whole class can become a simple data struct.

Reply:
In the current plan, it is very likely that the bias goes into the sensor calibration in the end, and the time-varying bias is modeled with the interpolator design. So this ``ImuState`` class will very likely be useless when we add the rigs. 

This being said, it is also logical to first make the rig design available and then adapt this PR accordingly : )